application/aam-backend-service/src/main/kotlin/com/aamdigital/aambackendservice/changes/core/CouchDatabaseChangeDetection.kt

```diff
@@ -0,0 +1,94 @@
+package com.aamdigital.aambackendservice.changes.core
+
+import com.aamdigital.aambackendservice.changes.core.event.DatabaseChangeEvent
+import com.aamdigital.aambackendservice.changes.di.ChangesQueueConfiguration.Companion.DB_CHANGES_QUEUE
+import com.aamdigital.aambackendservice.changes.repository.SyncEntry
+import com.aamdigital.aambackendservice.changes.repository.SyncRepository
+import com.aamdigital.aambackendservice.couchdb.core.CouchDbStorage
+import com.aamdigital.aambackendservice.couchdb.core.getEmptyQueryParams
+import org.slf4j.LoggerFactory
+import reactor.core.publisher.Mono
+import java.util.*
+
+class CouchDatabaseChangeDetection(
+    private val couchDbStorage: CouchDbStorage,
+    private val documentChangeEventPublisher: ChangeEventPublisher,
+    private val syncRepository: SyncRepository,
+) : DatabaseChangeDetection {
+    private val logger = LoggerFactory.getLogger(javaClass)
+
+    companion object {
+        private val LATEST_REFS: MutableMap<String, String> = Collections.synchronizedMap(hashMapOf())
+        private const val CHANGES_LIMIT: Int = 100
+    }
+
+    /**
+     * Will reach out to CouchDb and convert _changes to  Domain.DocumentChangeEvent's
+     */
+    override fun checkForChanges(): Mono<Unit> {
+        logger.trace("[CouchDatabaseChangeDetection] start couchdb change detection...")
+        return couchDbStorage.allDatabases().flatMap { databases ->
+            val requests = databases.filter { !it.startsWith("_") }.map { database ->
+                fetchChangesForDatabase(database)
+            }
+            Mono.zip(requests) {
+                it.map { }
+            }
+        }.map {
+            logger.trace("[CouchDatabaseChangeDetection] ...completed couchdb change detection.")
+        }
+    }
+
+    private fun fetchChangesForDatabase(database: String): Mono<Unit> {
+        logger.trace("[CouchDatabaseChangeDetection] check changes for database \"{}\"...", database)
+
+        return syncRepository.findByDatabase(database).defaultIfEmpty(SyncEntry(database = database, latestRef = ""))
```

Comment:
does this do multiple requests if multiple paginated results?

Reply:
Good point. Currently we fetch all changes at once, could be improved with pagination. I will create an improvement issue for that.

Also a good improvement: configure "start-sync-at-timestamp" or "start-sync-on-latest" configuration. Maybe also default behaviour for first-time starts to use the latest revs on all dbs.

Reply:
created issues: https://github.com/Aam-Digital/aam-services/issues/3 and https://github.com/Aam-Digital/aam-services/issues/4